storage/points_writer.go

```diff
@@ -10,3 +11,62 @@ import (
 type PointsWriter interface {
 	WritePoints(context.Context, []models.Point) error
 }
+
+type BufferedPointsWriter struct {
+	sync.Mutex
+	buf []models.Point
+	n   int
+	wr  PointsWriter
+	err error
+}
+
+func NewBufferedPointsWriter(size int, pointswriter PointsWriter) *BufferedPointsWriter {
+	return &BufferedPointsWriter{
+		buf: make([]models.Point, size),
+		wr:  pointswriter,
+	}
+}
+
+// WritePoints writes the points to the underlying PointsWriter.
+func (b *BufferedPointsWriter) WritePoints(ctx context.Context, p []models.Point) error {
+	for len(p) > b.Available() && b.err == nil {
+		var n int
+		if b.Buffered() == 0 {
+			// Large write, empty buffer.
+			// Write directly from p to avoid copy.
+			b.err = b.wr.WritePoints(ctx, p)
```

Comment:
I expect you want to add a `break` here or set `n` to `len(p)` so that this does not loop indefinitely or until WritePoints returns an error.